core/src/kura.rs

```diff
@@ -180,86 +190,156 @@ impl Kura {
     }
 
-    /// Put a block in the queue to be stored by Kura. If the queue is
-    /// full, wait until there is space.
-    pub async fn store_block_async(&self, block: VersionedCommittedBlock) {
-        if let Err(SendError(block)) = self.block_sender.send(block).await {
-            error!(?block, "unable to write block, kura thread was closed");
+    /// Search through blocks for the height of the block with the given hash.
+    pub fn get_block_height_by_hash(&self, hash: &HashOf<VersionedCommittedBlock>) -> Option<u64> {
+        let hash_data_guard = self.block_data.lock();
+
+        (*hash != Hash::zeroed().typed())
+            .then_some(())
+            .and(
+                hash_data_guard
+                    .iter()
+                    .position(|(block_hash, _block_arc)| block_hash == hash),
+            )
+            .map(|index| index as u64 + 1)
+    }
+
+    /// Get a reference to block by height, loading it from disk if needed.
+    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    pub fn get_block_by_height(&self, block_height: u64) -> Option<Arc<VersionedCommittedBlock>> {
+        let mut data_array_guard = self.block_data.lock();
+        if block_height == 0 || block_height > data_array_guard.len() as u64 {
+            return None;
+        }
+        let block_number: usize = (block_height - 1)
+            .try_into()
+            .expect("Failed to cast to u32.");
+
+        #[allow(clippy::option_if_let_else)]
+        if let Some(block_arc) = data_array_guard[block_number].1.as_ref() {
+            Some(Arc::clone(block_arc))
+        } else {
+            let block_store = self.block_store.lock();
+            let (block_start, block_len) = block_store
+                .read_block_index(block_number as u64)
+                .expect("Failed to read block index from disk.");
+
+            let mut block_buf =
+                vec![0_u8; usize::try_from(block_len).expect("index_len didn't fit in 32-bits")];
+            block_store
+                .read_block_data(block_start, &mut block_buf)
+                .expect("Failed to read block data.");
+            let block = VersionedCommittedBlock::decode_versioned(&block_buf)
+                .expect("Failed to decode block");
+
+            let block_arc = Arc::new(block);
+            data_array_guard[block_number].1 = Some(Arc::clone(&block_arc));
+            Some(block_arc)
         }
     }
 
-    /// Put a block in the queue to be stored by Kura. If the queue is
-    /// full, block the current thread until there is space.
+    /// Get a reference to block by hash, loading it from disk if needed.
+    ///
+    /// Internally this function searches linearly for the block's height and
+    /// then calls `get_block_by_height`. If you know the height of the block
+    /// call it instead.
+    pub fn get_block_by_hash(
+        &self,
+        block_hash: &HashOf<VersionedCommittedBlock>,
+    ) -> Option<Arc<VersionedCommittedBlock>> {
```

Comment:
to me it seems semantically incorrect to return `Arc` instead of `&` if `Kura` is the single source of truth and `VersionedCommittedBlock` is `Sync`

Reply:
The reason we return Arc is that I plan for Kura to unload unused blocks in order to keep memory usage down. Arc allows us to do that safely because we can see how many owners there are and drop if Kura is the only one.